### ISSUE TITLE
Sporadic C# compilation issue

### ISSUE THREAD

**Issue by fabiocav** [Azure/azure-functions-host#999]
Over the past few days, a few function apps have run into a compilation failure that seems to be caused by an attempt to queue additional Roslyn compilation events after the compilation is marked as completed, leading to an `InvalidOperationException` with the message: `Cannot enqueue data after PromiseNotToEnqueue.`

Opening this for investigation and tracking.

**Comment by paulbatum** [comment]
Looks like a Roslyn issue, we are following up with them.

**Comment by paulbatum** [comment]
@fabiocav any updates on this? If we don't have a date for when this is likely to be unblocked I think I should move it out of the January milestone.

**Comment by fabiocav** [comment]
Still an ongoing discussion with the team. It's not likely to be unblocked in this milestone, but it sounds like there has been some work done to address this issue in 2.0.0 (Roslyn)

**Comment by DamianFlynn** [comment]
This is still happening, three different function apps, and different functions within each app are now presenting the same problem as covered here

**Comment by fabiocav** [comment]
@DamianFlynn thank you for reporting. Unfortunately, a restart is indeed the way to address this at the moment, but we understand this is not ideal. We'll be working on updating the compiler references soon, now that 2.0 has been released, both, to address issues like this and bring C# 7.0 support.

**Comment by tomasmcguinness** [comment]
This issue occurred in a Queue bound function yesterday, nearly ruining a demo we were giving to a customer. The function in question was deployed weeks ago and just suddenly stopped working. As we're using the consumption model, I didn't find an obvious way to restart them as suggested by @DamianFlynn. I just disabled and reenabled the function, which seems to have solved the issue.  Do you have an ETA on when this will be resolved?

**Comment by mamaso** [comment]
Ran into this here as well: http://stackoverflow.com/questions/43085082/azure-function-with-servicebustrigger-stops-receiving-messages?noredirect=1#comment73298766_43085082

**Comment by DamianFlynn** [comment]
I am currently converting my solution to a class library. Will this approach circumvent the problem until such time as 2.0 is live?

Thanks

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mathewc** [comment]
Assigning this back to triage so we can get traction on this - customers are hitting this, and we're also seeing this in our runner from time to time, generating sev 2 incidents.

**Comment by gibwar** [comment]
Just ran in to this issue today as well on an HTTP triggered function. Restarting the app service brought everything back up. Also on consumption plan.

@tomasmcguinness: you can restart the service by going to the function app like normal, then function app settings, and under the "Manage" section click on "Go to App Service Settings" and from there you can stop and start it like a normal app service.

**Comment by fabiocav** [comment]
Reopening for triage as this did not get resolved as expected

**Comment by paulbatum** [comment]
We need a workaround so the app doesn't end up in a permanently stuck state when this occurs.

**Comment by paulbatum** [comment]
@fabiocav added retries and additional logging in attempt to address this, see https://github.com/Azure/azure-webjobs-sdk-script/commit/f24f85cd1c2f0ad4a5f2c1bc12f86faeb8b13d1a

This change should be deployed globally by 7/21. If it reproduces beyond that we should reactivate.

**Comment by OShevnin-Quest** [comment]
Was the change deployed on 7/21? I've got the similar error with my function today

**Comment by rwaal** [comment]
I seem to have run in this problem today, while editing a functions app via the Azure portal. It's the first time I've experienced this problem. Like others also mentioned, a restart of the Web application resolved the problem.

**Comment by tlmii** [comment]
I just ran into this for the first time. It appears to have started sometime between Friday afternoon and this morning. Like all the others, the restart worked fine for me, though is obviously not ideal. 

Mine was an HttpTrigger and consumption plan. I hope it is obvious how this is problematic - things just stop working without changes by us and without warning.

Any updates on this?

**Comment by fabiocav** [comment]
We've reached out to the compiler team as we were still seeing instances of this happening. We've put some mitigation steps that help _some_ scenarios, but as you found, we're not as aggressive with those steps. 
We did get an update from the team and there's a fix in the latest compiler version. We'll be deploying that with our next release.

**Comment by rs38** [comment]
I ran into this issue today with an .csx Script Function. Does a precompiled deployment fix this?

**Comment by fabiocav** [comment]
@rs38 yes, it does, as there's no runtime compilation in those scenarios. This is being addressed in this sprint.

**Comment by MeshTucker** [comment]
I can confirm that we are seeing this issue on our eventhub triggered functions.

We saw it last week on our consumption plan function, and again today on another consumption plan function.

**Comment by fabiocav** [comment]
@MeshTucker, fixes to address this are being rolled out and should be fully deployed by Wednesday. You can subscribe to this issue to receive updates: https://github.com/Azure/app-service-announcements/issues/39

**Comment by fashaikh** [comment]
@fabiocav : I hit this today, when visiting the portal. Function just stopped working without making any changes to it. Restart worked. Waiting on this fix to be deployed fully.

**Comment by RocketPop** [comment]
FYI I just got this today when editing a function in the portal.

**Comment by aronbardsley** [comment]
I just got this today. Refer to screenshot below. I really need this fixed as we are relying on functions in production. This issue caused a reporting system to go down - I've not changed the function in a month or two. 

![error](https://user-images.githubusercontent.com/16513722/32251325-a9992b82-bee4-11e7-92af-7ca7559c2b50.png)

**Comment by tangozulu32123** [comment]
Error:
Function ($ThinkSpeakEventHubTrigger) Error: C# compilation service error: Cannot enqueue data after PromiseNotToEnqueue. Cannot enqueue data after PromiseNotToEnqueue. 
Session Id: fc64551eee7f4f608f6cfcaa80c9b9be
Timestamp: 2018-01-08T05:22:07.897Z

**Comment by tangozulu32123** [comment]
Error:
Function () Error: C# compilation service error: Cannot enqueue data after PromiseNotToEnqueue. Cannot enqueue data after PromiseNotToEnqueue. 
Session Id: fc64551eee7f4f608f6cfcaa80c9b9be
Timestamp: 2018-01-08T05:22:07.897Z

**Comment by tangozulu32123** [comment]
Error:
Function () Error: C# compilation service error: Cannot enqueue data after PromiseNotToEnqueue. Cannot enqueue data after PromiseNotToEnqueue. 
Session Id: 
Timestamp: 2018-01-08T05:22:07.897Z

**Comment by fabiocav** [comment]
Thank you @tangozulu32123. This is assigned to the current sprint and we hope to have additional fixes and mitigation steps for this issue soon.

**Comment by fabiocav** [comment]
Updating milestone to better reflect when this will be done.

**Comment by sisoje** [comment]
same here, restart was needed

**Comment by snbcdev** [comment]
Ran into this issue over the weekend, our function app stopped working on 7/28/18 at 11:51 PM Pacific Time. When I logged into Azure Portal and looked at the app's code, I saw the erorr notification with the message "Cannot enqueue data after PromiseNotToEnqueue." Restarting the app fixed it.

**Comment by mathewc** [comment]
Guys - we're looking at this now and will get a fix out for this in the next release.

**Comment by paulbatum** [comment]
Until the fix is deployed globally, the best workaround is to switch to writing your function as a class library in VS or VS code (e.g. https://docs.microsoft.com/en-us/azure/azure-functions/functions-develop-vs) so that no compilation is required on the server side. I understand that moving from csx to cs is not 100% trivial and it might not be viable, but I would say that we are still at least a week away from having this fix be deployed globally so you might want to consider it. Sorry I dont have a better workaround to share right now.

**Comment by IvanovOleg** [comment]
This issue still exist!

**Comment by gridr-ch** [comment]
Hi @paulbatum ,
Is there an update / fix to this issue?

**Comment by paulbatum** [comment]
It seems that despite multiple attempts to address this issue through roslyn updates and other mitigations, this error can still occur. I don't have any updates to share at this time. The same recommendation that I made before still applies - you can completely avoid impact by deploying compiled binaries. Our dev tools in VS and VS Code make this straightforward.

**Comment by paulbatum** [comment]
We are planning on taking another Roslyn update (see the issue referenced above).

**Comment by Hazhzeng** [comment]
Another customer is raising the same issue 183252697.

**Comment by paulbatum** [comment]
The Roslyn update when live in May 2020 and as far as I'm aware there have been no further reports of the issue since then, closing.

**Comment by ConnorMcMahon** [comment]
We saw another issue of this appear on the latest version of V3 on an internal support case.
